bcgov/biohubbc
# BHBC-2300: Apply 'Viewer' and 'Editor' Permissions to Project / Survey Views

# Overview

As a Project Lead, I want to make sure that people I’ve invited to my team can only see the functions their role allows them to so they aren’t getting confused as to why some functionality appears broken.

## Links to Jira tickets

 - https://quartech.atlassian.net/browse/BHBC-2300

## Description of relevant changes

 - Created an API method to get the user's current roles for the given project
 - Created a Project Auth State context to determine the auth state for project and survey views
 - Created a component guard to hide UI elements based on accepted project roles
 - Created a route guard to protect app routes based on accepted project roles. Current protected routes include:
   - Project Manage Users page
   - Edit Project page
   - Create Survey page
   - Edit Survey page

## Testing Procedures

0. Create a BCeID Basic in the BCeID Dev environment. You can do this here: https://www.development.bceid.ca/register/.
1. From your IDIR account, add your BCeID user to a project as a **‘Viewer’.**
2. Go to Manage Users. Ensure that your newly added BCeID user does not have admin permissions by changing their role to _Creator._
3. Open a new Incognito (Chrome)/Private Browsing (Firefox)/InPrivate (Edge) window **(Ctrl + Shift + N)**. Sign into SIMS as your BCeID user.
4. View the project as the 'Viewer' user. It should:

 - [x] NOT show the following interactive components:
    - [x] Add Survey
    - [x] Upload Documents
    - [x] Delete Document (see individual document row items)
    - [x] Edit Location
  - [x] NOT be able to click **Project Settings**

5. From the project view, change your browser's URL from `/details` to `/edit`. It should:

 - [x] Show the Forbidden page.

6. From the project view, change your browser's URL from `/details` to `/users`. It should:

 - [x] Show the Forbidden page.

7. Navigate to a survey on the same project. It should:

 - [x] NOT show the following interactive components:
    - [x] Import Observations
    - [x] Delete Observations
    - [x] Import Summary Results
    - [x] Delete Summary Results
    - [x] Upload Documents
    - [x] Delete Document (see individual document row items)
    - [x] Edit Study Area
  - [x] NOT allow you to click **Survey Settings**

8. From the survey view, change your browser's URL from `/details` to `/edit`. It should:

 - [x] Show the Forbidden page.

9. From the survey view, change your browser's URL from `/{surveyId}/details` to `/create`. It should:

 - [x] Show the Forbidden page.

10. Change the project role of your BCeID user to be an **‘Editor’.**
11. View the project as the 'Editor' user. It should:

 - [x] NOT show the following project settings options:
    - Delete Project
    - Manage Project Team

12. From the project view, change your browser's URL from `/details` to `/users`. It should:

 - [x] Show the Forbidden page.
